Gems/Terrain/Code/Tests/TerrainBulkQueryTests.cpp

```diff
@@ -0,0 +1,938 @@
+    *  to a lack of proper thread safety. However, it's also possible that optimizations to the different queries could accidentally
+    *  produce different results as well, so it's good to have this safety net here.
+    */
+
+    class TerrainBulkQueryTest
+        : public TerrainTestFixture
+    {
+    protected:
+
+        // Use the ProcessHeightsFromRegion API as our baseline that we'll compare the other query APIs against.
+        void GenerateBaselineHeightData(
+            const AZ::Aabb& inputQueryRegion,
+            const AZ::Vector2& inputQueryStepSize,
+            AzFramework::Terrain::TerrainDataRequests::Sampler sampler,
+            AZStd::vector<AZ::Vector3>& queryPositions,
+            AZStd::vector<AZ::Vector3>& resultPositions)
+        {
+            queryPositions.clear();
+            resultPositions.clear();
+
+            auto perPositionCallback = [&queryPositions, &resultPositions](
+                                              [[maybe_unused]] size_t xIndex, [[maybe_unused]] size_t yIndex,
+                                              const AzFramework::SurfaceData::SurfacePoint& surfacePoint, bool terrainExists)
+            {
+                queryPositions.emplace_back(surfacePoint.m_position.GetX(), surfacePoint.m_position.GetY(), 0.0f);
+                resultPositions.emplace_back(surfacePoint.m_position);
+
+                // For these unit tests, we expect every point queried to have valid terrain data.
+                EXPECT_TRUE(terrainExists);
+            };
+
+            AzFramework::Terrain::TerrainDataRequestBus::Broadcast(
+                &AzFramework::Terrain::TerrainDataRequests::ProcessHeightsFromRegion, inputQueryRegion, inputQueryStepSize,
+                perPositionCallback, sampler);
+
+            EXPECT_EQ(queryPositions.size(), resultPositions.size());
+        }
+
+        // Compare two sets of output data and verify that they match.
+        void ComparePositionData(const AZStd::vector<AZ::Vector3>& baselineValues, const AZStd::vector<AZ::Vector3>& comparisonValues)
+        {
+            // Verify that we have the same quantity of results in both sets.
+            ASSERT_EQ(baselineValues.size(), comparisonValues.size());
+
+            // Verify that every value is found exactly once in each set. The two sets might not have the values in the same order though,
+            // so we need to search for each value, verify it's found, and verify that it hadn't previously been found.
+            AZStd::vector<bool> matchFound(baselineValues.size(), false);
+            for (auto& comparisonValue : comparisonValues)
+            {
+                auto foundValue = AZStd::find(baselineValues.begin(), baselineValues.end(), comparisonValue);
+                EXPECT_NE(foundValue, baselineValues.end());
+                if (foundValue != baselineValues.end())
+                {
+                    EXPECT_FALSE(matchFound[foundValue - baselineValues.begin()]);
```

Comment:
nit: consider making `foundValue - baselineValues.begin()` a separate variable like `valueIndex` to make this a little more clear.

Reply:
looks like you actually did this in at least one place below 😛 best to make it consistent.

Reply:
Fixed!